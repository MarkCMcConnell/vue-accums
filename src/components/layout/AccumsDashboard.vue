<template>
  <section>
    <h3>My Balances</h3>
    <DataTable :data="formatAccumulators" :fields="accumulatorFields"/>
  </section>
</template>

<script>
// General UI Components
import DataTable from '../ui/DataTable.vue'
// Mixins
import { dataHelpers } from '../../mixins/dataHelpers.js'

export default {
  components: {
    DataTable
  },
  props: {
    accumsData: {
      type: Array,
      required: true
    }
  },
  mixins: [dataHelpers],
  data () {
    return {
      accumulatorFields: [
        {
          label: 'Name',
          field: 'Name'
        },
        {
          label: 'Type',
          field: 'IndividualAccum'
        },
        {
          label: 'Coverage',
          field: 'CoverageType'
        },
        {
          label: 'Plan Year',
          field: 'PlanYear'
        },
        {
          label: 'Amount Met',
          field: 'CurrentAmount'
        },
        {
          label: 'Max Amount',
          field: 'MaximumAmount'
        },
        {
          label: 'Percent Met',
          field: 'ProgressBar'
        }
      ]
    }
  },
  computed: {
    formatAccumulators () {
      const accums = this.accumsData.filter(accum => {
        return accum.SpecificType !== 'GC'
      })
      // Array that will be pushed to the DataTables component
      let dataValues = []

      for (let obj of accums) {
        // Object that will hold the new values to be pushed to the array
        let data = {}

        for (let field of this.accumulatorFields) {
          let currentKey = obj[field.field]

          if (field.field === 'IndividualAccum') {
            if (currentKey) {
              currentKey = 'Individual'
            } else {
              currentKey = 'Family'
            }
          }

          if (field.field === 'CoverageType') {
            // Use this switch statement to customize or add values
            switch (currentKey) {
              case 'M':
                currentKey = 'Medical'
                break
              case 'D':
                currentKey = 'Dental'
                break
              default:
                currentKey = 'Unknown'
                break
            }
          }

          if (field.field === 'CurrentAmount') {
            currentKey = this.formatCurrency(obj.CurrentAmount)
          }

          if (field.field === 'MaximumAmount') {
            currentKey = this.formatCurrency(obj.MaximumAmount)
          }

          if (field.field === 'ProgressBar') {
            currentKey = this.calculatePercent(
              obj.CurrentAmount,
              obj.MaximumAmount
            )
          }
          // Avoid setting an empty row
          if (currentKey !== '' && currentKey !== undefined) {
            data[field.field] = currentKey
          }
          // Push to the array that will be passed to the coverages table
        }
        dataValues.push(data)
      }
      return dataValues
    }
  }
}
</script>

<style>
</style>
