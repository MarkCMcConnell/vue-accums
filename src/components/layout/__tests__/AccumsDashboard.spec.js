/* eslint-disable */
import { createLocalVue, shallowMount } from '@vue/test-utils'

import AccumsDashboard from '../AccumsDashboard.vue'

const localVue = createLocalVue()

describe('AccumsDashboard', () => {
	let wrapper

	beforeEach(() => {
		wrapper = shallowMount(AccumsDashboard, {
			localVue,
			propsData: {
				accumsData: [{
					Name: 'Dental',
					IndividualAccum: true,
					CoverageType: 'D',
					PlanYear: 2019,
					CurrentAmount: 1,
					MaximumAmount: 1000,
					AddDateTime: '\/Date(1528978573953-0400)\/'
				}]
			}
		})
	})

	afterEach(() => {
		wrapper.destroy()
	})

	it('renders the component', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})

	describe('formatAccumulators', () => {
		it('converts IndividualAccum to Individual or Family', () => {
			expect(wrapper.vm.formatAccumulators).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						IndividualAccum: 'Individual'
					})
				])
			)
		})

		it('converts coverage types to full words', () => {
			expect(wrapper.vm.formatAccumulators).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						CoverageType: 'Dental'
					})
				])
			)
		})

		it('converts CurrentAmount to currency format', () => {
			expect(wrapper.vm.formatAccumulators).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						CurrentAmount: '$1.00'
					})
				])
			)
		})

		it('converts MaximumAmount to currency format', () => {
			expect(wrapper.vm.formatAccumulators).toEqual(
				expect.arrayContaining([
					expect.objectContaining({
						MaximumAmount: '$1000.00'
					})
				])
			)
		})
	})
})