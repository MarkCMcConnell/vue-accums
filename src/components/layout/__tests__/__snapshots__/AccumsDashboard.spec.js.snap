// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccumsDashboard renders the component 1`] = `
<section>
  <h3>My Balances</h3>
  <datatable-stub fields="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]" data="[object Object]"></datatable-stub>
</section>
`;
